src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/NodeMonitoringConnectionPlugin.java

```diff
@@ -136,48 +132,44 @@ public Object execute(Class<?> methodInvokeOn, String methodName, Callable<?> ex
     initMonitorService();
 
     Object result;
-    ExecutorService executor = null;
+    MonitorConnectionContext monitorContext = null;
+
     try {
       this.log.logTrace(String.format(
           "[NodeMonitoringConnectionPlugin.execute]: method=%s.%s, monitoring is activated",
               methodInvokeOn.getName(), methodName));
 
-      checkNewConnection(this.proxy.getCurrentConnection());
+      this.checkIfChanged(this.currentConnectionProvider.getCurrentConnection());
 
-      this.monitorContext = this.monitorService.startMonitoring(
+      monitorContext = this.monitorService.startMonitoring(
+          this.connection, //abort current connection if needed
           this.nodeKeys,
-          this.proxy.getCurrentHostInfo(),
+          this.currentConnectionProvider.getCurrentHostInfo(),
           this.propertySet,
           failureDetectionTimeMillis,
           failureDetectionIntervalMillis,
           failureDetectionCount);
 
-      executor = Executors.newSingleThreadExecutor();
-      final Future<Object> executeFuncFuture = executor.submit(() -> this.nextPlugin.execute(methodInvokeOn, methodName, executeSqlFunc));
-      executor.shutdown(); // stop executor to accept new tasks
-
-      boolean isDone = executeFuncFuture.isDone();
-      while (!isDone) {
-        TimeUnit.MILLISECONDS.sleep(CHECK_INTERVAL_MILLIS);
-        isDone = executeFuncFuture.isDone();
-
-        if (this.monitorContext.isNodeUnhealthy()) {
-          //throw new SocketTimeoutException("Read time out");
-          throw new CJCommunicationsException("Node is unavailable.");
-        }
-      }
+      result = this.nextPlugin.execute(methodInvokeOn, methodName, executeSqlFunc);
 
-      result = executeFuncFuture.get();
-    } catch (ExecutionException exception) {
-      final Throwable throwable = exception.getCause();
-      if (throwable instanceof Error) {
-        throw (Error) throwable;
+    } catch (SQLException | CJException exception) {
+      throw exception;
+    } catch (Exception ex) {
+      if (monitorContext != null && monitorContext.isNodeUnhealthy()) {
```

Comment:
We don't need the `catch (Exception ex)` clause if we add the checks to finally.